### ISSUE TITLE
Calling of Subaddresses/ small additional features

### ISSUE THREAD

**Issue by Luxovious** [RobTillaart/PCA9634#8]
Hi Rob,
I have been playing around with the library a little longer to check wether some of the functionallity does not work.

#### Allcall-Address works fine
I have found a very little issue regarding the allcall address. You have stated that it is 0x11 in the PCA9634 header file. However for this chip it is 0x70. You have put it under the comment "// not implemented yet", however it does not need any implementation. It works perfectly fine, when you create another instance of the chip with the allcall-addresse (e.g.: PCA9634 ledArray(0x70);)

#### Subaddresses do not work
Furthermore, I tried to use the subaddresses (1-3). I'm pretty sure that I activated them successfully. Unfortunately, I'm not able to communicate with the chip through this address. The datasheet mentions that when subaddresses are used, the chip does not send an ACK. Therefore, your comment "//not implemented yet" is correct. Do you intend to implement this feature some time later? I assume, that this is a lot of work, since you have to work at the core of your library. At my current project I would love the benefit of subaddresses, since I'm using 15 - 25 chips and I would like to supdivide them into two subadresses for a major part of the functions.

#### Implementation of software reset
In some instances it might be usefull to send a software reset to some/all chips. Currently, I do that manually, if required. I assume, some users might find this usefull, when debugging their software. In some applications I do a software reset before configuring the chip, since I then know that no old configuration is safed that might interfere. This is usefull when playing around with new code :-)

The datasheed states:

> **Software Reset**
The Software Reset Call (SWRST Call) allows all the devices in the I2C-bus to be reset to the power-up state value through a specific formatted I2C-bus command. To be performed correctly, it implies that the I2C-bus is functional and that there is no device hanging the bus.
The SWRST Call function is defined as the following:
>**1.** A START command is sent by the I2C-bus master.
>**2.** The reserved SWRST I2C-bus address ‘0000 011’ with the R/W bit set to ‘0’ (write) is sent by the I2C-bus master.
>**3.** The PCA9634 device(s) acknowledge(s) after seeing the SWRST Call address ‘0000 0110’ (06h) only. If the R/W bit is set to ‘1’ (read), no acknowledge is returned to the I2C-bus master.
>**4.** Once the SWRST Call address has been sent and acknowledged, the master sends 2 bytes with 2 specific values (SWRST data byte 1 and byte 2):
 >     - Byte 1 = A5h: the PCA9634 acknowledges this value only. If byte 1 is not equal to A5h, the PCA9634 does not acknowledge it.
 >     - Byte 2 = 5Ah: the PCA9634 acknowledges this value only. If byte 2 is not equal to 5Ah, then the PCA9634 does not acknowledge it. If more than 2 bytes of data are sent, the PCA9634 does not acknowledge any more.
>**5.** Once the right 2 bytes (SWRST data byte 1 and byte 2 only) have been sent and correctly acknowledged, the master sends a STOP command to end the SWRST Call: the PCA9634 then resets to the default value (power-up value) and is ready to be addressed again within the specified bus free time (tBUF). The I2C-bus master must interpret a non-acknowledge from the PCA9634 (at any time) as a ‘SWRST Call Abort’. The PCA9634 does not initiate a reset of its registers. This happens only when the format of the SWRST Call sequence is not correct.

I hope you get better soon,
best regards Johannes

**Comment by RobTillaart** [comment]
#### Sub-addresses do not work

Not worked with sub addresses (mostly trying to catch up my projects backlog as that has priority)

I saw these bit flags when looking at the other issue
```
#define PCA9634_MODE1_SUB1          0x08  // 0 = disable      1 = enable
#define PCA9634_MODE1_SUB2          0x04  // 0 = disable      1 = enable
#define PCA9634_MODE1_SUB3          0x02  // 0 = disable      1 = enable
```

These are not set by default and might be needed to get them to work..



#### Implementation of software reset

This looks like a function "bigger" than this library although the PCA9634 can handle it.
Is that correct? 
And if so, should it be part of this library?
Note: it might (will) affect other devices on the I2C bus, causing unexpected behavior.

If it should be part of this library, I propose to split this issue in two new issues (copy paste txt)
1. sub call and all call
2. sw reset

Which of these two has your highest priority?

**Comment by Luxovious** [comment]
Sub-addresses have a much higher priority to me than software reset.

In a worst case scenario, a power off sets even a completely bricked PCA9634 in it's working default mode. At a later stage I want to run around 20 of the chips for illumination my parents-in-law's garden 24-7. Therefore, a weekly (or user requested) SW reset might be an interesting future feature.

- Regarding the sub-addresses, they should be activated in the Mode1 register. With setMode1 and getMode1 I can verify, that this works.
```cpp
ledArray.setMode1(0x19);  // 25 in decimal, SLEEP, SUB1, ALLCALL activated
uint8_t test2 = ledArray.getMode1();
Serial.print("Mode1 Register: ");
Serial.println(test2);
```
 --> Serial print: "Mode1 Register: 25"
- However, due to the checks in the writeMode/readMode functions (line 113ff in .cpp, only `(reg == PCA9634_MODE1) || (reg == PCA9634_MODE2)`) I cannot change the address of the sub-address to verify if that works.
- I played around with this and changed the checks. Reading the byte of the controll register (`#define PCA9634_SUBADR(x)           (0x0E +(x))  // x = 1..3`, line 59 .h) indicates, that I'm able to read the default address. Iassume, that I would also write to it and therefore change the address
- As a first test, I created another instance (?) of the PCA9634 class `PCA9634 ALLled(0x70);` in the same code. It gives the option to use the allcall address (0x70) without any flaws next to the individual address in the same code. Can be used right out of the box with your library (I used it as a check for debugging in some of my tests).
- Creating another instance `PCA9634 SUB1LED(0x71);` with the activeted default sub-address1 (0x71) **does not lead to a result.**

**Comment by RobTillaart** [comment]
Created a separate issue #10 for the software reset so we can focus on the SUB CALL / ALL CALL here.

**Comment by Luxovious** [comment]
An interesting thing that I found was that I have to initialise the AllCall address just as I have to initialise the individual address. Even the the chip has been configured before. I assume I'm still too unfamiliar with the I2C bus and you consider that as perfectly normal behaviour. Anyways, just as documentation: this code works even after having used the code with 

```cpp
PCA9634 ledArray(0x40);
PCA9634 ALLled(0x70);
ALLled.begin();
ALLled.setLedDriverMode(1, PCA9634_LEDPWM);
ALLled.write1(1,120);
```

(updated code to a block of cpp code, gives syntax highlighting - check with edit )

**Comment by Luxovious** [comment]
Thanks for the edit. Didn't know this feature :-)

**Comment by RobTillaart** [comment]
especially handy with longer blocks of code, 
there are more languages supported, just use the typical extension e.g.  js  or java or python

**Comment by RobTillaart** [comment]
The goal to set is to have something like this working

```cpp
PCA9634 ledArray0(0x40);
...
PCA9634 ledArrayF(0x4F);

PCA9634 sub_call_one(0x71);      //  ==   0xE2  default sub call address.

sub_call_one.begin();
sub_call_one.setLedDriverMode(1, PCA9634_LEDPWM);     //   all sub call 1 enabled  are expected to react. 
sub_call_one.write1(1, 120);
```

To keep it manageable we might need some functions like:

```cpp
void enableSubCall(uint8_t nr);   //  nr = { 1, 2, 3 }
void disableSubCall(uint8_t nr);   //  nr = { 1, 2, 3 }
bool isEnabledSubCall(uint8_t nr);   //  nr = { 1, 2, 3 }

ledArrayF.enableSubCall(1);
```

**Comment by RobTillaart** [comment]
can you patch the library

```cpp
void PCA9634::reset()
{
  _data = 0;
  _error = 0;

  uint8_t mode1_mask = PCA9634_MODE1_AUTOINCR2 | PCA9634_MODE1_ALLCALL;
  uint8_t mode1_mask |= PCA9634_MODE1_SUB1;   //  <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< add this line 
  writeReg(PCA9634_MODE1, mode1_mask);  //  AUTOINCR | NOSLEEP | ALLADRR  0x81
}
```

and then something like this?
```cpp
PCA9634 ledArray0(0x40);
PCA9634 sub_call_one(0x71);      //  ==   0xE2  default sub call address.

sub_call_one.begin();
sub_call_one.setLedDriverMode(1, PCA9634_LEDPWM);     //   all sub call 1 enabled  are expected to react. 
sub_call_one.write1(1, 120);
```

After a first (re)read of the datasheet I do not see a real technical difference between the SUB CALL and the ALL CALL in terms of how the mechanism works. You set a flag to react on a 2nd or a 3rd I2C address besides your "own" address.

( I call it a day, to be continued later this week)

**Comment by Luxovious** [comment]
I will do so during the week. Unfortunately, not today as I'm busy tonight.

**Comment by Luxovious** [comment]
I might have find the problem(s) why I wasn't able to make it work.

A `PCA9634.begin()` function calls the reset funktoin. Anytime this is done (for example when initializing the ALLCALL or SUBCALL) the configuration of the chip(s) is changed. Therefore, I removed the reset function from of the begin function and used it seperately - only once in the beginning.
```cpp
  ledArray.begin();
  ledArray.reset();
```
I just commented the reset-function

```cpp
bool PCA9634::begin()
{
  _wire->begin();
  if (! isConnected()) return false;
  //reset();
  return true;
}
```
Frankly, it also should not be called a "reset" and rather a pre-configure (or similar) function. As the reset function works entirely different. In detail, this function is also slightly incoherent with the TDS (albeit it works perfectly):
```cpp
void PCA9634::reset()
{
  _data = 0;
  _error = 0;
  uint8_t mode1_mask = PCA9634_MODE1_AUTOINCR2 | PCA9634_MODE1_ALLCALL;
  uint8_t mode1_mask |= PCA9634_MODE1_SUB1;   //  <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< add this line 
  writeReg(PCA9634_MODE1, mode1_mask);  //  AUTOINCR | NOSLEEP | ALLADRR  0x81
}
```
The
```cpp
uint8_t mode1_mask = PCA9634_MODE1_AUTOINCR2
```
doesnot make a difference, as this bit is read only. The auto-increment is turned on by default. However, it is deactivated by adressing
```cpp
#define PCA9634_MODE1               0x00;
```
as it was used in the current version of the library. Addressing it with
```cpp
#define PCA9634_MODE1               0x80;
```
would activate the auto-increment just as it is used in in your (perfect!) use of
```cpp
#define PCA9634_PWM(x)              (0x82+(x))    // Auto-Increment for all registers.
```

Now at this point it still does not work. After "reseting" the chip in the beginning, the MODE1 register reads 0x01, however it should read 0x09. The Sub-Address 1 is not activated. However, when activating the address manually:
```cpp
  ledArray.setMode1(0x09);  // 9 in decimal, NOSLEEP, SUB1, ALLCALL activated
```
the chip does what it should.
```cpp
  sub_call_one.setLedDriverMode(4, PCA9634_LEDPWM);     //   all sub call 1 enabled  are expected to react. 
  sub_call_one.write1(4, 120);
```

So my final suggestions: the chips should only be configured once in the beginning. The reset-function could be replaced with a "configure" function and later the software reset function (#10) could be used to restart bricked chips without pwr off. Finally, somewhere in the readme, it should be noted, that allcall/subcall should only be created (`PCA9634 ALLCALL`) but not configured (`.begin()`).

**Comment by RobTillaart** [comment]
Thanks for these insights, sounds like a "breaking release", will try to catch up next week.

**Comment by RobTillaart** [comment]
Found some time today. Created a develop branch to prepare 0.2.0 version.

This branch implements the following functions to enable/disable SUB and ALL calls.
Also renamed **reset()** to **initial()** in this branch - if you have a better name let me know. 

```cpp
  //  SUB CALL SUPPORT  (since 0.2.0)
  //  nr = { 1, 2, 3 }
  void enableSubCall(uint8_t nr);
  void disableSubCall(uint8_t nr);
  bool isEnabledSubCall(uint8_t nr);

  //  ALL CALL SUPPORT  (since 0.2.0)
  void enableAllCall();
  void disableAllCall();
  bool isEnabledAllCall();
```

Above you mentioned you used 

```cpp
ledArray.setMode1(0x09);  // 9 in decimal, NOSLEEP, SUB1, ALLCALL activated
```

Which would now become more explicit / readable
```cpp
ledArray.enableSubCall(1);
```

Please give it a try

**Comment by RobTillaart** [comment]
@Luxovious 
Added functions and updated readme.md to discuss the making of groups with **ALL CALL and SUB CALL**.

The list of new functions for 0.2.0 is now:
```cpp
  //  nr = { 1, 2, 3 }
  bool    enableSubCall(uint8_t nr);
  bool    disableSubCall(uint8_t nr);
  bool    isEnabledSubCall(uint8_t nr);
  bool    setSubCallAddress(uint8_t nr, uint8_t address);
  uint8_t getSubCallAddress(uint8_t nr);

  bool    enableAllCall();
  bool    disableAllCall();
  bool    isEnabledAllCall();
  bool    setAllCallAddress(uint8_t address);
  uint8_t getAllCallAddress();
```

@Luxovious 
Can you check if the readme.md section is readable (to get started).

**Comment by Luxovious** [comment]
Hi @RobTillaart,
I have read through the readme-file and here are my remarks. I will go through the code as well and discuss it afterwards.

In the section "Constructor" you wrote about the "reset" function. I believe this should be changed. In your post above you mentioned that you renamed it to inital. However, I believe this might be missleading and **config** would be the most appropriate name for what the function does. Initial(ization) would indicate, that this function is required to be used once at the start. In contrast, if you want to change to configuration of the PCA9634/5 chips during the runtime of the program, the wording might be missleading. "Config" or similar would describe that possibility more accurate.

In some cases a few words of explanation might help the unexperienced user to better go along with your library.
- LEDDriverMode: add sub-headline "Configure LED behaviour"
- Read and write: add sub-headline "Read and write individual values to LED channels. Requires LEDs' DriverMode of the specific channels to be in PWM mode"
- Mode registers: add sub-headline "Used to configure the PCA963x general behaviour"
- SubCall & All Call: AllCall is automatically activated for each device on startup. This should be mentioned in this chapter

Spelling mistakes or similar:
- `uint8_t writeMode(uint8_t reg, uint8_t value) configuration of one of the two configuration registers. check datasheet for details.`  <-- **Check** instead of **check**
- Table: spell our "RO" so it reads "read only". I allready got that part because I know that these are read only.

I'll work on the code later on.

**Comment by Luxovious** [comment]
One thing I had to double check:

Chapter "**Group PWM and frequency**"
```cpp
void setGroupFREQ(value); // see datasheet for details.
uint8_t getGroupFREQ(); // returns the freq of the PWM group.
```
--> GroupFREQ is used for **blinking** the group of configured LED. It's an 8-bit value from 0 to 256 with each value representing a linear increase of around 41 ms. So 0x00 results in a 41 ms blinking period (on AND off) and 0xFF in 10.73 s.

**Comment by RobTillaart** [comment]
Fixed all of the above in documentation and pushed into develop branch.

**Comment by Luxovious** [comment]
Hi Rob, I have read the source code and I found a few things. 

New configure function:

```cpp
void PCA9634::configure()
{
  _data = 0;
  _error = 0;

  uint8_t mode1_mask = PCA9634_MODE1_AUTOINCR2 | PCA9634_MODE1_ALLCALL;
  writeReg(PCA9634_MODE1, mode1_mask);  //  AUTOINCR | NOSLEEP | ALLADRR  0x81
}
```

When using writeReg on the mode 1 or mode 2 register (addresses 0x00 or 0x01) Autoincrement CANNOT be set. In the mode 1 register the thre MSB are "read-only". Autoincrement is set by adding 0x80 to the address of the mode 1 (mode 2 / PWM) register. Using the addresses 0x00 or 0x01 would turn auto-increment off. To be fair, I'm currently wondering if this is an issue or even necessary.

I haven't tried the code yet, as I'm still working through it. Maybe it might be wise to wait until I have given my feedback (in pieces) before changing the code, to save some time on your side :-)

**Comment by RobTillaart** [comment]
You've mentioned the ```autoincrement``` read only attribute before. 
It has no function in the code so it should be removed, but it escaped my attention (more than once).
(now removed in my local repo, no push yet)

**Comment by Luxovious** [comment]
PCA9634.h:
```cpp
  // TODO PWM also in %% ?
  void     setGroupPWM(uint8_t value) { writeReg(PCA9634_GRPPWM, value); };
  uint8_t  getGroupPWM() { return readReg(PCA9634_GRPPWM); };

  // TODO set time in milliseconds and round to nearest value?
  void     setGroupFREQ(uint8_t value) { writeReg(PCA9634_GRPFREQ, value); };
  uint8_t  getGroupFREQ() { return readReg(PCA9634_GRPFREQ); };
```
I added the ";" at the end of each line in the header file. Maybe it is wrong, however it's the same way in all other lines, so I assumed it should also be here.

**Comment by Luxovious** [comment]
I have checked the subcall-functions.
Results: activation and deactivation works perfectly. However, I have not yet checked these functions:
```cpp
  //  nr = { 1, 2, 3 }
  bool    isEnabledSubCall(uint8_t nr);
  bool    setSubCallAddress(uint8_t nr, uint8_t address);
  uint8_t getSubCallAddress(uint8_t nr);

  bool    enableAllCall();
  bool    disableAllCall();
  bool    isEnabledAllCall();
  bool    setAllCallAddress(uint8_t address);
  uint8_t getAllCallAddress();
```
I will do that today in the evening or at the end of the week (business trip in between).

**Comment by RobTillaart** [comment]
> I added the ";" at the end of each line in the header file. Maybe it is wrong, however it's the same way in all other lines, so I assumed it should also be here.

IIRC  if the function has a body it is not needed, if it is only a function signature it is mandatory.

**Comment by Luxovious** [comment]
So I just finished testing. Your subcall/allcall functions all work fine in every extend that I tested.
Thank you very much for all the time and work you put into the library.

**Comment by RobTillaart** [comment]
I will merge asap

**Comment by RobTillaart** [comment]
Merged, release 0.2.0 published (CI build is running)
